### ISSUE TITLE
A few fixes/improvements maybe. Sorry no pull request.

### ISSUE THREAD

**Issue by ghogan42** [nyxkrage/ebook-groupchat#1]
Here are a couple of small issues I notice going through this code. But none of this was worth a pull request I figure. Sorry for the wall of text 😬 

I'm attaching my llama-cpp-python version of the script.
I'm also attaching the epub script in case you want to see how I batch processed a pile of epubs.
[Scipts_and_Sample_json.zip](https://github.com/nyxkrage/ebook-groupchat/files/15387682/Scipts_and_Sample_json.zip)

## Problem 1

 In your `def mark_speakers(chapter, characters):`
You have this line

```
{", ".join(characters)}
```
That seems to break the character list by turning it into something like this: 

```
1, .,  , L, e, s, l, i, e, 
, 2, .,  , J, a, m, e, s,  , D, o, u, g, l, a, s, 
, 3, .,  , A, n, d, r, e, w, 
, 4, .,  , A, l, a, n, 
```
Instead of this:
```
1. Leslie
2. James Douglas
3. Andrew
4. Alan
```

## Problem 1 fix

Probably just change that line to get rid of the `.join`
```
{characters}
```` 
like I think you did to the other functions.
This is the biggest problem with the code right now since only gpt-4 doesn't choke on this part. And if you fix it, local models do a much better job on these tasks.

## Problem 2
This isn't much of a problem really.
But in `def clarify_character(conversation, characters):`

you have this text in the samples you show the llm 
```
...
8. The blond man, the lanky fellow
9. Two uniformed cops (not individually named)

The character list is given in the following format
1. Name, alias
2. Name, alias

Which character is the following referring to, use only the characters as in the list.
```

But for the actual generation instruction you just have this

```
{characters}

Which character is the following referring to, use only the characters as in the list.
```

## Problem 2 fix

Probably just add in the missing lines "1. Name, alias" lines so the real instruction matches the sample instructions.
```
{characters}

The character list is given in the following format
1. Name, alias
2. Name, alias

Which character is the following referring to, use only the characters as in the list.
```

## These are not problems maybe but just things I noticed

1. Most local models write extra text before outputting the character list. So I strip it with regex at the bottom of `get_characters()` using
```
    characters = re.sub(r".*?(1\..*)", r"\1", characters, flags=re.DOTALL)
```
2. In the `clarify_characters()` function a lot of local models apologize instead of just outputting a name as requested. I changed my instruction to the following because I don't have a better idea. But I'm sure there is a better/shorter way to phrase it.
```
Which character is the following referring to, use only the characters as in the list. Use their full name if possible and no other title or description. Your response must be either a single (full) name or "Unknown". No other text or numbers.
```
3. Early on, I had an error on lines that did not contain a ":" in `interrogate_speakers()`. I had to add an extra `if-statement` to skip lines without a colon, but maybe this should not have been necessary? So mine looks like this at the top...
```
def interrogate_speakers(marked_text, characters):
    lines = marked_text.splitlines()
    conversation = []
    for i, line in enumerate(lines):
        if ": " not in line:
            continue  # Skip lines that do not contain ": "
        messages = [
...
```

**Comment by nyxkrage** [comment]
Hey :wave:
Thanks for catching these, can't believe I missed that first one. I've implemented all but your final suggestion, as I feel like there not being a colon in the lines is more indicative of something having gone wrong with the previous prompting and would rather have it error.

**Comment by ghogan42** [comment]
No problem. I'm not trying to make you change anything. Just trying to run the script and see how well it could do with the models I have already downloaded.

And I figured you would not want to change the colon thing. I just had to change it when I rant the sctipt for the first time or I couldn't run it all the way through. It probbaly would run properly now since the {charaters} thing was probably all that was making the local models screw up and not be able to find the current character actor.
